control-plane/cni/main.go

```diff
@@ -135,56 +162,120 @@ func cmdAdd(args *skel.CmdArgs) error {
 
 	client, err := kubernetes.NewForConfig(restConfig)
 	if err != nil {
-		return fmt.Errorf("Error initializing Kubernetes client: %s", err)
+		return fmt.Errorf("error initializing Kubernetes client: %s", err)
 	}
 
 	pod, err := client.CoreV1().Pods(podNamespace).Get(ctx, podName, metav1.GetOptions{})
 	if err != nil {
 		return err
 	}
 
-	// TODO: Add tests for all of the logic below
-	// TODO: Remove has hasBeenInjected and instead look for consul.hashicorp.com/cni-proxy-config annotation
-	// TODO: Add wait and timeout for annotations to show up
-	if hasBeenInjected(*pod) {
-		// If everything is good, add an annotation to the pod
-		// TODO: Remove this as it is just a stub to prove that we can do kubernetes things with the plugin
-		annotations := map[string]string{
-			keyCNIStatus: "true",
-		}
-		pod.SetAnnotations(annotations)
-		_, err = client.CoreV1().Pods(podNamespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
-		if err != nil {
-			return err
-		}
-	} else {
-		logger.Debug("skipping traffic redirect on un-injected pod")
+	// skip injecting pod that has consul.hashicorp.com/connect-inject: false
+	if skipConnectInject(*pod) {
+		logger.Debug("skipping traffic redirect on un-injected pod: %s", pod.Name)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// check to see if the cni-proxy-config annotation exists and if not, wait, retry and backoff.
+	exists := waitForAnnotation(*pod, annotationCNIProxyConfig, uint64(retries))
+	if !exists {
+		logger.Error("could not retrieve annotation: %s on pod: %s", annotationCNIProxyConfig, pod.Name)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// parse the cni-proxy-config annotation into an iptables.Config object.
+	iptablesCfg, err := parseAnnotation(*pod, annotationCNIProxyConfig)
+	if err != nil {
+		logger.Error("could not parse annotation: %s, error: %v", annotationCNIProxyConfig, err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// get the DNS IP from the environment.
+	dnsIP := searchDNSIPFromEnvironment(*pod, cfg.DNSPrefix)
+	if dnsIP != "" {
+		logger.Error("assigned consul dns ip to %s", dnsIP)
+		iptablesCfg.ConsulDNSIP = dnsIP
 	}
 
-	// TODO: Get transparent proxy annotations and merge with proxy config
-	// TODO: Redirect traffic :)
+	// apply the iptables rules.
+	err = iptables.Setup(iptablesCfg)
+	if err != nil {
+		logger.Error("could not apply iptables setup: %v", err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
 
-	// Pass through the result for the next plugin
+	logger.Debug("traffic redirect rules applied to pod: %s", pod.Name)
+	// Pass through the result for the next plugin even though we are the final plugin in the chain.
```

Comment:
Yes, because I make it so in the file watcher. It covers the following scenarios:

- If a config file does not exist yet (ie our installer daemonset if the first started), then we will wait for the config file to show up.
- If a config file exists, append our config.
- If the config file changes, validate it to make sure our config is the last in the chain and fix if need be.